File: backend-v2/internal/common/encryption/decrypt_strategy.go
package encryption

type DecryptStrategy interface {
	Decrypt(ciphertext string, key []byte, additionalData []byte) (string, error)
}

type StandardDecrypt struct {
	cipher *Cipher
}

func NewStandardDecrypt(cipher *Cipher) *StandardDecrypt {
	return &StandardDecrypt{cipher: cipher}
}

func (s *StandardDecrypt) Decrypt(ciphertext string, key []byte, additionalData []byte) (string, error) {
	return s.cipher.Decrypt(ciphertext, key, additionalData)
}

type FallbackDecrypt struct {
	cipher *Cipher
}

func NewFallbackDecrypt(cipher *Cipher) *FallbackDecrypt {
	return &FallbackDecrypt{cipher: cipher}
}

func (f *FallbackDecrypt) Decrypt(ciphertext string, key []byte, additionalData []byte) (string, error) {
	plaintext, err := f.cipher.Decrypt(ciphertext, key, additionalData)
	if err != nil && additionalData != nil {
		return f.cipher.Decrypt(ciphertext, key, nil)
	}
	return plaintext, err
}

File: backend/src/models/utils/decryptStrategy.js
const AAD_ENFORCEMENT_MESSAGE =
  'Decryption failed — ciphertext lacks AAD binding and AAD migration enforcement is active. ' +
  'Re-save integration to migrate.'

class FallbackDecrypt {
  constructor(cipher, enforceAadBinding = false) {
    this.cipher = cipher
    this.enforceAadBinding = enforceAadBinding
  }

  decrypt(ciphertext, key, additionalData) {
    try {
      return this.cipher.decrypt(ciphertext, key, additionalData)
    } catch (error) {
      if (additionalData !== null && additionalData !== undefined) {
        if (this.enforceAadBinding) {
          throw new Error(`${AAD_ENFORCEMENT_MESSAGE} Original: ${error.message}`)
        }
        console.warn(
          '[SECURITY] AAD fallback triggered — data encrypted without AAD binding. ' +
            'Re-save integration entry to migrate to AAD-protected encryption.',
        )
        return this.cipher.decrypt(ciphertext, key, null)
      }
      throw error
    }
  }
}

export {FallbackDecrypt}
